### DISCUSSION TITLE
Question on rendering!

### DISCUSSION THREAD

**Discussion by capslocknanda** [ocornut/imgui#6414] [New Users Build/Link/Run issues ONLY!]
Thank you for sharing this awesome work.
I am rendering this figure using gl4 and GLFW and glad,
![image](https://user-images.githubusercontent.com/46029468/236757347-350ca84e-45e9-4a97-be9f-ef1699e8b57b.png)
and here is my main loop to render this:
int main()
{
    // glfw: initialize and configure
    // ------------------------------
    glfwInit();
    if (!glfwInit())
    {
        // Handle initialization error
        std::cout << "Failed to initialize GLFW" << std::endl;
    }

    glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
    glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 6);
    glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);

#ifdef __APPLE__
    glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
#endif

    // glfw window creation
    // --------------------
    GLFWwindow *window = glfwCreateWindow(800, 600, "drawing lines", NULL, NULL);
    // glfwSetWindowAttrib(window, GLFW_MAXIMIZED, GLFW_TRUE);
    if (!window)
    {
        std::cout << "Failed to create GLFW window" << std::endl;
        glfwTerminate();
        return -1;
    }
    glfwMakeContextCurrent(window);

    // glad: load all OpenGL function pointers
    // ---------------------------------------

    if (!gladLoadGLLoader((GLADloadproc)glfwGetProcAddress))
    {
        std::cout << "Failed to initialize GLAD" << std::endl;
        return -1;
    }

    camera.position = glm::vec3(3, 3, 3);

    // 3d lines example
    Line line1(glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(1.0f, 0.0f, 0.0f));
    line1.setColor(glm::vec3(1.0f, 0.0f, 0.0f));
    Line line2(glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0));
    line2.setColor(glm::vec3(0.0f, 1.0f, 0.0f));
    Line line3(glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 0.0f, 1.0f));
    line3.setColor(glm::vec3(0.0f, 0.0f, 1.0f));

    std::vector<std::vector<int>> edgesList{{0, 1}, {1, 2}, {1, 5}, {2, 3}, {3, 4}, {5, 6}, {6, 7}, {1, 8}, {1, 9}};

    glm::mat4 projection = glm::perspective(glm::radians(45.0f), (float)SCR_WIDTH / (float)SCR_HEIGHT, 0.1f, 100.0f);

    float angle = 0.0f;
    float rotationSpeed = 50.0f;

    std::vector<std::vector<double>> examplePose{
        {0.564655185, 0},            // head
        {0.512931049, 0.25469169},   // neck
        {0.0517241396, 0.209115282}, // left hip
        {0, 0.648793578},            // right hip
        {0.637931049, 0.705093801},  // left shoulder
        {1, 0.276139408},            // left elbow
        {0.883620679, 0.675603211},  // left wrist
        {0.232758626, 0.571045578},  // right shoulder
        {0.198275864, 0.975871325},  // right elbow
        {0.711206913, 1},            // right wrist
    };

    // render loop
    // -----------
    while (!glfwWindowShouldClose(window))
    {
        float currentFrame = glfwGetTime();
        deltaTime = currentFrame - lastFrame;
        // input
        // -----
        processInput(window);

        // render
        // ------
        glClearColor(0.0, 0.0, 0.0, 1.0);
        glClear(GL_COLOR_BUFFER_BIT);

        // angle = deltaTime * rotationSpeed;

        // update camera position (rotating)
        // camera.position = glm::vec3(3 * cos(glm::radians(angle)), 3, 3 * sin(glm::radians(angle)));
        glm::mat4 view = lookAt(camera.position, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f));

        for (auto &edge : edgesList)
        {
            int v1 = edge[0];
            int v2 = edge[1];
            Line lineN(glm::vec3(-examplePose[v1][0], -examplePose[v1][1], 0),
                       glm::vec3(-examplePose[v2][0], -examplePose[v2][1], 0));
            lineN.setColor(glm::vec3(1, 0, 0));
            lineN.setMVP(projection * view);
            lineN.draw();
        }

        line1.setMVP(projection * view);
        line2.setMVP(projection * view);
        line3.setMVP(projection * view);

        line1.draw();
        line2.draw();
        line3.draw();

        // line4.draw();

        // glfw: swap buffers and poll IO events (keys pressed/released, mouse moved etc.)
        // -------------------------------------------------------------------------------
        glfwSwapBuffers(window);
        glfwPollEvents();
    }

    // glfw: terminate, clearing all previously allocated GLFW resources.
    // ------------------------------------------------------------------
    glfwTerminate();
    return 0;
}

How can I render the same in a sub-window in another of my program which uses Imgui with SDL and opengl,
I am currently doing this :

void render3DPose(cv::Mat &PoseKeypoint, std::vector<std::vector<int>> edgesList, GLuint rbo,
                  GLuint framebuffer, GLuint texture) {
    ImGui::Begin("Rendering 3D Pose");
    ImGui::SetWindowPos(ImVec2(380, 140));
    ImVec2 windowSize = ImVec2(450, 550);
    ImVec2 size = ImVec2(ImGui::GetWindowWidth(), ImGui::GetWindowHeight());
    ImGui::SetWindowSize("Rendering 3D Pose", windowSize);

    glViewport(0, 0, (int)windowSize.x, (int)windowSize.y);
    glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
    glClearColor(0.2f, 0.3f, 0.4f, 1.0f);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    float angle = 40.0f;
    Camera camera;
    camera.position = glm::vec3(3.0f, 3.0f, 3.0f);
    glm::mat4 projection =
        glm::perspective(glm::radians(45.0f), (float)SCR_WIDTH / (float)SCR_HEIGHT, 0.1f, 100.0f);

    glm::mat4 view =
        glm::lookAt(camera.position, glm::vec3(0.0f, 0.0f, 0.0f), glm::vec3(0.0f, 1.0f, 0.0f));

    cv::Mat PoseKeypoint3dmat = PoseKeypoint.clone();
    //// Normalize keypoint coordinates to 0-1 range
    cv::normalize(PoseKeypoint3dmat.col(0), PoseKeypoint3dmat.col(0), 0.0, 1.0, cv::NORM_MINMAX);
    cv::normalize(PoseKeypoint3dmat.col(1), PoseKeypoint3dmat.col(1), 0.0, 1.0, cv::NORM_MINMAX);

    for (auto &edge : edgesList) {
        int v1 = edge[0];
        int v2 = edge[1];
        Line lineN(
            glm::vec3(PoseKeypoint3dmat.at<float>(v1, 0), -PoseKeypoint3dmat.at<float>(v1, 1), 0),
            glm::vec3(PoseKeypoint3dmat.at<float>(v2, 0), -PoseKeypoint3dmat.at<float>(v2, 1), 0));
        lineN.setColor(glm::vec3(1, 0, 0));
        lineN.setMVP(projection * view);
        lineN.draw();
    }

    Line line1(glm::vec3(0, 0, 0), glm::vec3(1, 0, 0));
    line1.setColor(glm::vec3(1, 0, 0));
    Line line2(glm::vec3(0, 0, 0), glm::vec3(0, 1, 0));
    line2.setColor(glm::vec3(0, 1, 0));
    Line line3(glm::vec3(0, 0, 0), glm::vec3(0, 0, 1));
    line3.setColor(glm::vec3(0, 0, 1));

    line1.setMVP(projection * view);
    line2.setMVP(projection * view);
    line3.setMVP(projection * view);

    line1.draw();
    line2.draw();
    line3.draw();

    glBindFramebuffer(GL_FRAMEBUFFER, 0);

    // Clear the screen
    glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

    ImTextureID tex_id = (void *)(intptr_t)texture;
    ImGui::Image(tex_id, windowSize, ImVec2(0, 1), ImVec2(1, 0));

    ImGui::End();
}
I am using the same glm camera parameters, but ending up getting this:

![image](https://user-images.githubusercontent.com/46029468/236758302-2cae4f5b-1b33-4b6e-af29-347b279d4c81.png)

Any help will be appreciated. Thank you!

**Comment by ocornut** [comment]
I am sorry but nothing in your question is specific to dear imgui, this is the wrong forum to ask this.

**Reply by capslocknanda** [reply]
I did convert it to a ImTextureID. My question is is there any way to render another sdl window in side ImGui subwindow?
Thank you!

**Reply by ocornut** [reply]
No, you can render a texture meaning you have to render-to-texture using opengl to do this, as you did. Your problem seems to be a 3d/projection transform problem.
